x/did/keeper/msg_server.go

```diff
@@ -313,48 +173,53 @@ func (k msgServer) SetVerificationRelationships(
 	goCtx context.Context,
 	msg *types.MsgSetVerificationRelationships,
 ) (*types.MsgSetVerificationRelationshipsResponse, error) {
-	ctx := sdk.UnwrapSDKContext(goCtx)
-	k.Logger(ctx).Info("request to set verification relationships", "target did", msg.Id, "method", msg.MethodId)
-	// retrieve the did document
-	didDoc, found := k.Keeper.GetDidDocument(ctx, []byte(msg.Id))
-	if !found {
-		err := sdkerrors.Wrapf(types.ErrDidDocumentNotFound, "did document at %s not found", msg.Id)
-		k.Logger(ctx).Error("request to set verification relationships failed", "error", err.Error())
-		return nil, err
-	}
-	// any verification method in the authentication relationship can update the DID document
-	if !didDoc.HasRelationship(types.NewBlockchainAccountID(ctx.ChainID(), msg.Signer), types.Authentication) {
-		err := sdkerrors.Wrapf(
-			types.ErrUnauthorized,
-			"signer %s not authorized to set verification relationships on the target did document at %s",
-			msg.Signer, msg.Id,
-		)
-		k.Logger(ctx).Error("request to set verification relationships failed", "error", err.Error())
+
+	if err := executeOnDidWithRelationships(
+		goCtx, &k.Keeper,
+		newConstraints(types.Authentication),
+		msg.Id, msg.Signer,
+		func(didDoc *types.DidDocument) error {
+			return didDoc.SetVerificationRelationships(msg.MethodId, msg.Relationships...)
+		}); err != nil {
 		return nil, err
 	}
 
-	// set the verification relationships
-	err := didDoc.SetVerificationRelationships(msg.MethodId, msg.Relationships...)
-	if err != nil {
-		k.Logger(ctx).Error("request to set verification relationships failed", "error", err.Error(), "did", didDoc.Id)
+	return &types.MsgSetVerificationRelationshipsResponse{}, nil
+}
+
+// AddController add a new controller to a DID
+func (k msgServer) AddController(
+	goCtx context.Context,
+	msg *types.MsgAddController,
+) (*types.MsgAddControllerResponse, error) {
+	if err := executeOnDidWithRelationships(
+		goCtx, &k.Keeper,
+		newConstraints(types.Authentication),
+		msg.Id, msg.Signer,
+		func(didDoc *types.DidDocument) error {
+			return didDoc.AddControllers(msg.ControllerDid)
+		}); err != nil {
 		return nil, err
 	}
 
-	// persist the did document
-	k.Keeper.SetDidDocument(ctx, []byte(msg.Id), didDoc)
-	k.Logger(ctx).Info("Set verification relationship from did document for", "did", msg.Id, "controller", msg.Signer)
+	return &types.MsgAddControllerResponse{}, nil
+}
 
-	// update the Metadata
-	if err := updateDidMetadata(&k.Keeper, ctx, didDoc.Id); err != nil {
-		k.Logger(ctx).Error("request to update did document metadata failed", "error", err.Error(), "did", didDoc.Id)
+// DeleteController remove an existing controller from a DID document
+func (k msgServer) DeleteController(
```

Comment:
Do we need a cli command that calls this function in the msg_server?

Reply:
yes we do!